// @cell4rom/front-line/esm5/lib/rs-pagination-lib.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
/**
 * #### Pagination within API RSET
 *
 * Author : romdhanisamir94\@gmail.com
 */
var RsPaginationLibComponent = /** @class */ (function () {
    function RsPaginationLibComponent(activatedRoute, router) {
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.sizes = [5, 10, 20];
    }
    /**
     * @return {?}
     */
    RsPaginationLibComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    RsPaginationLibComponent.prototype.lastPage = /**
     * @return {?}
     */
    function () {
        this.router.navigate([], {
            relativeTo: this.activatedRoute,
            queryParams: { page: this.pagesNumber },
            queryParamsHandling: "merge"
        });
    };
    /**
     * @return {?}
     */
    RsPaginationLibComponent.prototype.firstPage = /**
     * @return {?}
     */
    function () {
        this.router.navigate([], {
            relativeTo: this.activatedRoute,
            queryParams: { page: 1 },
            queryParamsHandling: "merge"
        });
    };
    /**
     * @return {?}
     */
    RsPaginationLibComponent.prototype.previousPage = /**
     * @return {?}
     */
    function () {
        this.router.navigate([], {
            relativeTo: this.activatedRoute,
            queryParams: { page: --this.page },
            queryParamsHandling: "merge"
        });
    };
    /**
     * @return {?}
     */
    RsPaginationLibComponent.prototype.nextPage = /**
     * @return {?}
     */
    function () {
        this.router.navigate([], {
            relativeTo: this.activatedRoute,
            queryParams: { page: ++this.page },
            queryParamsHandling: "merge"
        });
    };
    /**
     * @return {?}
     */
    RsPaginationLibComponent.prototype.changeSize = /**
     * @return {?}
     */
    function () {
        this.router.navigate([], {
            relativeTo: this.activatedRoute,
            queryParams: { size: this.currentSize },
            queryParamsHandling: "merge"
        });
    };
    RsPaginationLibComponent.decorators = [
        { type: Component, args: [{
                    selector: 'rs-pagination-api',
                    template: "\n  <div *ngIf=\"pagesNumber >= 1\">\n\t<div class=\"row\">\n\t\t\t<div class=\"col-md-12\">\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"firstPage()\" [disabled]=\"page === 1\"><i class=\"fa fa-angle-double-left\"></i></button>\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"previousPage()\" [disabled]=\"page == 1\"><i class=\"fa fa-angle-left\"></i></button>\n\t\t\t\t<span style=\"margin-left: 10px;margin-right: 10px;\">Page {{page}}/{{pagesNumber}}</span> \n\t\t\t\t<span>  </span>\n        <span style=\"margin-left: 10px;margin-right: 10px;\">Size</span>\n\t\t\t\t<select class=\"form-control form-control-sm pageSizeSelectBox\" id=\"currentSize\" name=\"currentSize\" [(ngModel)]=\"currentSize\" (change)=\"changeSize()\">\n\t\t\t\t\t<option *ngFor=\"let size of sizes\" value=\"{{size}}\">{{size}}</option>\n        </select>\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"nextPage()\" [disabled]=\"page === pagesNumber\"><i class=\"fa fa-angle-right\"></i></button>\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"lastPage()\" [disabled]=\"page === pagesNumber\"><i class=\"fa fa-angle-double-right\"></i></button>\n\t\t\t</div>\n\t</div>\n</div>\n  ",
                    styles: ["\n  .m-l-1em { \n    margin-left:1em; \n    width: 25px;\n  }\n  \n  .pageSizeSelectBox {\n    display: -webkit-inline-box;\n    width: 11%;\n    padding: 5px; \n  }\n  \n  "]
                }] }
    ];
    /** @nocollapse */
    RsPaginationLibComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: Router }
    ]; };
    RsPaginationLibComponent.propDecorators = {
        page: [{ type: Input }],
        pagesNumber: [{ type: Input }],
        currentSize: [{ type: Input }]
    };
    return RsPaginationLibComponent;
}());
export { RsPaginationLibComponent };
if (false) {
    /** @type {?} */
    RsPaginationLibComponent.prototype.page;
    /** @type {?} */
    RsPaginationLibComponent.prototype.pagesNumber;
    /** @type {?} */
    RsPaginationLibComponent.prototype.currentSize;
    /** @type {?} */
    RsPaginationLibComponent.prototype.sizes;
    /** @type {?} */
    RsPaginationLibComponent.prototype.activatedRoute;
    /** @type {?} */
    RsPaginationLibComponent.prototype.router;
}
