// @cell4rom/front-line/esm2015/lib/v2/rs-pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { paginate } from './rs-model';
/**
 * #### Pagination without API RSET
 *
 * Author : romdhanisamir94\@gmail.com
 */
export class RsPaginationComponent {
    constructor() {
        this.currentPage = 1;
        this.sizes = [5, 10, 20];
        this.changePage = new EventEmitter(true);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.items && this.items.length) {
            this.pagesNumber = Math.ceil(this.items.length / this.currentSize);
            this.setPage(this.currentPage);
        }
    }
    /**
     * @return {?}
     */
    lastPage() {
        this.currentPage = this.pagesNumber;
        this.setPage(this.currentPage);
    }
    /**
     * @return {?}
     */
    firstPage() {
        this.currentPage = 1;
        this.setPage(this.currentPage);
    }
    /**
     * @return {?}
     */
    previousPage() {
        this.currentPage = this.currentPage - 1;
        this.setPage(this.currentPage);
    }
    /**
     * @return {?}
     */
    nextPage() {
        this.currentPage = this.currentPage + 1;
        this.setPage(this.currentPage);
    }
    /**
     * @param {?} size
     * @return {?}
     */
    changeSize(size) {
        this.setSize(size);
    }
    /**
     * @param {?} page
     * @return {?}
     */
    setPage(page) {
        /** @type {?} */
        let pager = paginate(this.items.length, page, this.currentSize);
        this.currentPage = pager.currentPage;
        this.pagesNumber = pager.totalPages;
        /** @type {?} */
        var pageOfItems = this.items.slice(pager.startIndex, pager.endIndex + 1);
        this.changePage.emit(pageOfItems);
    }
    /**
     * @param {?} size
     * @return {?}
     */
    setSize(size) {
        /** @type {?} */
        let pager = paginate(this.items.length, this.currentPage, size);
        this.currentPage = pager.currentPage;
        this.pagesNumber = pager.totalPages;
        /** @type {?} */
        var pageOfItems = this.items.slice(pager.startIndex, pager.endIndex + 1);
        this.changePage.emit(pageOfItems);
    }
}
RsPaginationComponent.decorators = [
    { type: Component, args: [{
                selector: 'rs-pagination',
                template: `
  <div>
	<div class="row">
			<div class="col-md-12">
				<button class='btn btn-info btn-sm m-l-1em' (click)="firstPage()" [disabled]="currentPage === 1"><i class="fa fa-angle-double-left"></i></button>
				<button class='btn btn-info btn-sm m-l-1em' (click)="previousPage()" [disabled]="currentPage == 1"><i class="fa fa-angle-left"></i></button>
				<span style="margin-left: 10px;margin-right: 10px;">Page {{currentPage}}/{{pagesNumber}}</span> 
				<span>  </span>
        <span style="margin-left: 10px;margin-right: 10px;">Size</span>
        <select class="form-control form-control-sm pageSizeSelectBox" id="currentSize" 
                name="currentSize" [(ngModel)]="currentSize" (change)="changeSize(currentSize)">
					<option *ngFor="let size of sizes" value="{{size}}">{{size}}</option>
        </select>
				<button class='btn btn-info btn-sm m-l-1em' (click)="nextPage()" [disabled]="currentPage === pagesNumber"><i class="fa fa-angle-right"></i></button>
				<button class='btn btn-info btn-sm m-l-1em' (click)="lastPage()" [disabled]="currentPage === pagesNumber"><i class="fa fa-angle-double-right"></i></button>
			</div>
	</div>
</div>
  `,
                styles: [`
  .m-l-1em { 
    margin-left:1em; 
    width: 25px;
  }
  
  .pageSizeSelectBox {
    display: -webkit-inline-box;
    width: 11%;
    padding: 5px; 
  }
  
  `]
            }] }
];
/** @nocollapse */
RsPaginationComponent.ctorParameters = () => [];
RsPaginationComponent.propDecorators = {
    items: [{ type: Input }],
    currentSize: [{ type: Input }],
    changePage: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    RsPaginationComponent.prototype.items;
    /** @type {?} */
    RsPaginationComponent.prototype.currentSize;
    /** @type {?} */
    RsPaginationComponent.prototype.currentPage;
    /** @type {?} */
    RsPaginationComponent.prototype.pagesNumber;
    /** @type {?} */
    RsPaginationComponent.prototype.sizes;
    /** @type {?} */
    RsPaginationComponent.prototype.changePage;
}
