// @cell4rom/front-line/esm5/lib/v2/rs-pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { paginate } from './rs-model';
/**
 * #### Pagination without API RSET
 *
 * Author : romdhanisamir94\@gmail.com
 */
var RsPaginationComponent = /** @class */ (function () {
    function RsPaginationComponent() {
        this.currentPage = 1;
        this.sizes = [5, 10, 20];
        this.changePage = new EventEmitter(true);
    }
    /**
     * @return {?}
     */
    RsPaginationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    RsPaginationComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.items && this.items.length) {
            this.pagesNumber = Math.ceil(this.items.length / this.currentSize);
            this.setPage(this.currentPage);
        }
    };
    /**
     * @return {?}
     */
    RsPaginationComponent.prototype.lastPage = /**
     * @return {?}
     */
    function () {
        this.currentPage = this.pagesNumber;
        this.setPage(this.currentPage);
    };
    /**
     * @return {?}
     */
    RsPaginationComponent.prototype.firstPage = /**
     * @return {?}
     */
    function () {
        this.currentPage = 1;
        this.setPage(this.currentPage);
    };
    /**
     * @return {?}
     */
    RsPaginationComponent.prototype.previousPage = /**
     * @return {?}
     */
    function () {
        this.currentPage = this.currentPage - 1;
        this.setPage(this.currentPage);
    };
    /**
     * @return {?}
     */
    RsPaginationComponent.prototype.nextPage = /**
     * @return {?}
     */
    function () {
        this.currentPage = this.currentPage + 1;
        this.setPage(this.currentPage);
    };
    /**
     * @param {?} size
     * @return {?}
     */
    RsPaginationComponent.prototype.changeSize = /**
     * @param {?} size
     * @return {?}
     */
    function (size) {
        this.setSize(size);
    };
    /**
     * @param {?} page
     * @return {?}
     */
    RsPaginationComponent.prototype.setPage = /**
     * @param {?} page
     * @return {?}
     */
    function (page) {
        /** @type {?} */
        var pager = paginate(this.items.length, page, this.currentSize);
        this.currentPage = pager.currentPage;
        this.pagesNumber = pager.totalPages;
        /** @type {?} */
        var pageOfItems = this.items.slice(pager.startIndex, pager.endIndex + 1);
        this.changePage.emit(pageOfItems);
    };
    /**
     * @param {?} size
     * @return {?}
     */
    RsPaginationComponent.prototype.setSize = /**
     * @param {?} size
     * @return {?}
     */
    function (size) {
        /** @type {?} */
        var pager = paginate(this.items.length, this.currentPage, size);
        this.currentPage = pager.currentPage;
        this.pagesNumber = pager.totalPages;
        /** @type {?} */
        var pageOfItems = this.items.slice(pager.startIndex, pager.endIndex + 1);
        this.changePage.emit(pageOfItems);
    };
    RsPaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'rs-pagination',
                    template: "\n  <div>\n\t<div class=\"row\">\n\t\t\t<div class=\"col-md-12\">\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"firstPage()\" [disabled]=\"currentPage === 1\"><i class=\"fa fa-angle-double-left\"></i></button>\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"previousPage()\" [disabled]=\"currentPage == 1\"><i class=\"fa fa-angle-left\"></i></button>\n\t\t\t\t<span style=\"margin-left: 10px;margin-right: 10px;\">Page {{currentPage}}/{{pagesNumber}}</span> \n\t\t\t\t<span>  </span>\n        <span style=\"margin-left: 10px;margin-right: 10px;\">Size</span>\n        <select class=\"form-control form-control-sm pageSizeSelectBox\" id=\"currentSize\" \n                name=\"currentSize\" [(ngModel)]=\"currentSize\" (change)=\"changeSize(currentSize)\">\n\t\t\t\t\t<option *ngFor=\"let size of sizes\" value=\"{{size}}\">{{size}}</option>\n        </select>\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"nextPage()\" [disabled]=\"currentPage === pagesNumber\"><i class=\"fa fa-angle-right\"></i></button>\n\t\t\t\t<button class='btn btn-info btn-sm m-l-1em' (click)=\"lastPage()\" [disabled]=\"currentPage === pagesNumber\"><i class=\"fa fa-angle-double-right\"></i></button>\n\t\t\t</div>\n\t</div>\n</div>\n  ",
                    styles: ["\n  .m-l-1em { \n    margin-left:1em; \n    width: 25px;\n  }\n  \n  .pageSizeSelectBox {\n    display: -webkit-inline-box;\n    width: 11%;\n    padding: 5px; \n  }\n  \n  "]
                }] }
    ];
    /** @nocollapse */
    RsPaginationComponent.ctorParameters = function () { return []; };
    RsPaginationComponent.propDecorators = {
        items: [{ type: Input }],
        currentSize: [{ type: Input }],
        changePage: [{ type: Output }]
    };
    return RsPaginationComponent;
}());
export { RsPaginationComponent };
if (false) {
    /** @type {?} */
    RsPaginationComponent.prototype.items;
    /** @type {?} */
    RsPaginationComponent.prototype.currentSize;
    /** @type {?} */
    RsPaginationComponent.prototype.currentPage;
    /** @type {?} */
    RsPaginationComponent.prototype.pagesNumber;
    /** @type {?} */
    RsPaginationComponent.prototype.sizes;
    /** @type {?} */
    RsPaginationComponent.prototype.changePage;
}
